### PR TITLE
[PM-8280] email forwarders

### PR DESCRIPTION
**Pull request by audreyality**
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/PM-8280

## 📔 Objective

Add forwarder support to credential generator

- Extends the `UserStateSubject` with classifier & encryptor support
  - Move generalized rx helpers to common so they can be used by `UserStateSubject`
  - Introduce `ObjectKey` to support encryption to POJOs in state storage
- Extend `CredentialGeneratorService` with forwarder integration support:
  - Introduce `{ forwarder: IntegrationId }` algorithm Id to model service integration
  - Include forwarder integration metadata in algorithms lists
  - Introduce `ForwarderEngine` to invoke RPCs in response to generator requests.
  - Introduce `AlgorithmInfo` to model localized credential generation metadata.
- Add forwarder list and generation features to v2 generator components

## 🏁 Remaining tasks
- [x] Define forwarder subject keys
- [x] Add encryptor injection to credential generator service
- ~~Use import configuration for settings migration~~ split to [PM-13733](https://bitwarden.atlassian.net/browse/PM-13733)
- [x] Implement forwarder settings component(s)
- [x] Address TODOs
- [x] Unit test "classified" format support in `UserStateSubject`

## 🦕 System Evolution

Switching to `{ 'engine': EngineId }` for built-in generators could improve cohesion throughout the generator system. It also eliminates confusion between the `'username'` category and the `{ engine: 'username' }` engineId.

Promote `ObjectKey` to a class. Several of the methods in `UserStateSubject`, particularly those used before classification, properly belong in `ObjectKey`. `ObjectKey` should also provide an optional serializer for parity with its deserializer.

## 📸 Screenshots

image
image
image
image
image
image
image

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

[PM-13733]: https://bitwarden.atlassian.net/browse/PM-13733?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by djsmith85 — CHANGES_REQUESTED** [review]
@audreyality Quite a bit to go through and took longer than I'd hoped for. Really great work and only have some minor things

File: libs/common/src/tools/state/identity-state-constraint.ts [hunk 1]
```diff
@@ -9,16 +16,30 @@ const EMPTY_CONSTRAINTS = new Proxy<any>(Object.freeze({}), {
...
   adjust(state: State) {
     return state;
   }
+
   fix(state: State) {
     return state;
   }
 }
+
+export function unconstrained$<State extends object>(): Observable<SubjectConstraints<State>> {
```
[the remark this answers is not in the document]
Comment (audreyality) [reply libs/common/src/tools/state/identity-state-constraint.ts:40]: The case is correct. `state` is the variable and `State` is the variable's type.

[the remark this answers is not in the document]
Comment (audreyality) [reply libs/common/src/tools/state/identity-state-constraint.ts:40]: Don't function names follow `camelCase`?

File: libs/tools/generator/core/src/engine/forwarder.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+import { AccountRequest, ForwarderConfiguration } from "./forwarder-configuration";
+import { ForwarderContext } from "./forwarder-context";
+import { CreateForwardingAddressRpc, GetAccountIdRpc } from "./rpc";
+
+/** Generation algorithms that produce randomized email addresses */
+export class Forwarder implements CredentialGenerator<ApiSettings> {
+  /** Instantiates the email randomizer
+   *  @param random data source for random data
+   */
```
[the remark this answers is not in the document]
Comment (audreyality) [reply libs/tools/generator/core/src/engine/forwarder.ts]: Addressed by 13d54288ba2.

File: libs/tools/generator/core/src/engine/forwarder.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+} from "@bitwarden/common/tools/integration/rpc";
+import { GenerationRequest } from "@bitwarden/common/tools/types";
+
+import { CredentialGenerator, GeneratedCredential } from "../types";
+
+import { AccountRequest, ForwarderConfiguration } from "./forwarder-configuration";
+import { ForwarderContext } from "./forwarder-context";
+import { CreateForwardingAddressRpc, GetAccountIdRpc } from "./rpc";
+
+/** Generation algorithms that produce randomized email addresses */
```
[the remark this answers is not in the document]
Comment (audreyality) [reply libs/tools/generator/core/src/engine/forwarder.ts]: Addressed by 13d54288ba2.

File: libs/common/src/tools/state/identity-state-constraint.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (audreyality) [reply libs/common/src/tools/state/identity-state-constraint.ts:41]: Missing docs addressed by 13d54288ba2.

File: libs/common/src/tools/dependencies.ts [hunk 4]
```diff
@@ -45,6 +47,12 @@ export type UserDependency = {
   userId$: Observable<UserId>;
 };
 
+export type UserBound<K extends keyof any, T> = { [P in K]: T } & { userId: UserId };
+
+export type SingleUserEncryptorDependency = {
+  singleUserEncryptor$: Observable<UserBound<"encryptor", UserEncryptor>>;
+};
+
```
[the remark this answers is not in the document]
Comment (audreyality) [reply libs/common/src/tools/dependencies.ts]: Addressed by 13d54288ba2.

**Review by djsmith85 — APPROVED** [review]

**Pull request merged**